### PR TITLE
[InlineCost] Correct the default branch cost for the switch statement

### LINKED ISSUE #81723: #77856 causes a minor performance regression in stage2 clang builds
FWIW, it seems like this causes a minor performance regression in stage2 clang builds (https://llvm-compile-time-tracker.com/compare.php?from=5aec9392674572fa5a06283173a6a739742d261d&to=5932fcc47855fdd209784f38820422d2369b84b2&stat=instructions:u). Though code size does reduce as well.

_Originally posted by @nikic in https://github.com/llvm/llvm-project/issues/77856#issuecomment-1937851382_

### PR DESCRIPTION
**Pull request by dianqk**
Fixes #81723.

I use the following patch to find functions that are not inlined after #77856.

<details><summary>patch.diff</summary>
<p>

```diff
diff --git a/llvm/lib/Analysis/InlineCost.cpp b/llvm/lib/Analysis/InlineCost.cpp
index e55eaa55f8e9..e325d18ab0a8 100644
--- a/llvm/lib/Analysis/InlineCost.cpp
+++ b/llvm/lib/Analysis/InlineCost.cpp
@@ -29,6 +29,7 @@
 #include "llvm/Analysis/TargetTransformInfo.h"
 #include "llvm/Analysis/ValueTracking.h"
 #include "llvm/Config/llvm-config.h"
+#include "llvm/Demangle/Demangle.h"
 #include "llvm/IR/AssemblyAnnotationWriter.h"
 #include "llvm/IR/CallingConv.h"
 #include "llvm/IR/DataLayout.h"
@@ -575,6 +576,8 @@ class InlineCostCallAnalyzer final : public CallAnalyzer {
   // True if the cost-benefit-analysis-based inliner is enabled.
   const bool CostBenefitAnalysisEnabled;
 
+  int DefaultBranchCost = 0;
+
   /// Inlining cost measured in abstract units, accounts for all the
   /// instructions expected to be executed for a given function invocation.
   /// Instructions that are statically proven to be dead based on call-site
@@ -701,8 +704,11 @@ class InlineCostCallAnalyzer final : public CallAnalyzer {
 
   void onFinalizeSwitch(unsigned JumpTableSize, unsigned NumCaseCluster,
                         bool DefaultDestUndefined) override {
-    if (!DefaultDestUndefined)
+    if (!DefaultDestUndefined) {
+      DefaultBranchCost = std::clamp<int64_t>(DefaultBranchCost + 2 * InstrCost,
+                                              INT_MIN, INT_MAX);
       addCost(2 * InstrCost);
+    }
     // If suitable for a jump table, consider the cost for the table size and
     // branch to destination.
     // Maximum valid cost increased in this function.
@@ -1132,6 +1138,7 @@ public:
   virtual ~InlineCostCallAnalyzer() = default;
   int getThreshold() const { return Threshold; }
   int getCost() const { return Cost; }
+  int getDefaultBranchCost() const { return DefaultBranchCost; }
   int getStaticBonusApplied() const { return StaticBonusApplied; }
   std::optional<CostBenefitPair> getCostBenefitPair() { return CostBenefit; }
   bool wasDecidedByCostBenefit() const { return DecidedByCostBenefit; }
@@ -3072,6 +3079,23 @@ InlineCost llvm::getInlineCost(
                             GetAssumptionCache, GetBFI, PSI, ORE);
   InlineResult ShouldInline = CA.analyze();
 
+  if (CA.getCost() > CA.getThreshold() &&
+      (CA.getCost() - CA.getDefaultBranchCost() <= CA.getThreshold())) {
+    auto ModuleName = Callee->getParent()->getName();
+    auto *CallerName = llvm::itaniumDemangle(Call.getCaller()->getName());
+    auto *CalleeName = llvm::itaniumDemangle(Callee->getName());
+    errs() << "NOT Inlining ModuleName: " << ModuleName << " Caller: " << CallerName
+           << ", Callee: " << CalleeName << ", Cost: " << CA.getCost()
+           << ", Threshold: " << CA.getThreshold()
+           << ", DefaultBranchCost: " << CA.getDefaultBranchCost();
+    if (auto *SP = Callee->getSubprogram()) {
+       auto FileName = SP->getFilename();
+       unsigned Line = SP->getLine();
+       errs() << ", FileName: " << FileName << "#L" << Line;
+    }
+    errs() << "\n";
+  }
+
   LLVM_DEBUG(CA.dump());
 
   // Always make cost benefit based decision explicit.
```

</p>
</details> 

There are over 20,000 call sites that don't satisfy the inline condition. I tried to select 10 of them:

<details><summary>Details</summary>
<p>

- Callee: `DoLowering(llvm::Function&, llvm::GCStrategy&)` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/llvm/lib/CodeGen/GCRootLowering.cpp#L201
- Callee: `getFromRangeMetadata(llvm::Instruction*)` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/llvm/lib/Analysis/LazyValueInfo.cpp#L589
- Callee: `clang::comments::DeclInfo::fill()` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/clang/lib/AST/Comment.cpp#L203
- Callee: `clang::APValue::DestroyDataAndMakeUninit()` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/clang/lib/AST/APValue.cpp#L403
- Callee: `clang::targets::MipsTargetInfo::getISARev() const` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/clang/lib/Basic/Targets/Mips.cpp#L61
- Callee: `llvm::isLegalUTF8(unsigned char const*, int)` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/llvm/lib/Support/ConvertUTF.cpp#L397
- Callee: `llvm::yaml::Input::createHNodes(llvm::yaml::Node*)` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/llvm/lib/Support/YAMLTraits.cpp#L401
- Callee: `clang::Parser::ParseOpenACCDirective()` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/clang/lib/Parse/ParseOpenACC.cpp#L1119
- Callee: `clang::OMPClauseReader::readClause()` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/clang/lib/Serialization/ASTReader.cpp#L10263
- Callee: `clang::CodeGen::CodeGenFunction::EmitLandingPad()` FileName: https://github.com/llvm/llvm-project/blob/5aec9392674572fa5a06283173a6a739742d261d/clang/lib/CodeGen/CGException.cpp#L825

</p>
</details> 

There are complex switch statements that cannot be transformed to simpler structures.

The earliest commit of the related code is: https://github.com/llvm/llvm-project/commit/919f9e8d65ada6552b8b8a5ec12ea49db91c922a. I tried to understand the following code with https://github.com/llvm/llvm-project/pull/77856#issuecomment-1993499085.

https://github.com/llvm/llvm-project/blob/5932fcc47855fdd209784f38820422d2369b84b2/llvm/lib/Analysis/InlineCost.cpp#L709-L720

I think only scenarios where there is a default branch were considered.

Taking https://llvm.godbolt.org/z/5cno1TnGx as an example, we need additional compare and jump instructions when there is a default branch, otherwise we just need a jump instruction.

```asm
foo: # @foo
  cmp rdi, 6
  ja .LBB0_6
  jmp qword ptr [8*rdi + .LJTI0_0]
...
bar: # @bar
  jmp qword ptr [8*rdi + .LJTI1_0]
...
```

But I don't know why it's `4 * InstrCost` and not `3 * InstrCost`.

Taking https://llvm.godbolt.org/z/MEsf9sno7 as an example, we can reduce a set of compare and jump instructions when the number of branches is small.

```asm
foo: # @foo
  cmp rdi, 4
  je .LBB0_5
  cmp rdi, 2
  je .LBB0_4
  test rdi, rdi
  jne .LBB0_6
...
bar: # @bar
  cmp rdi, 4
  je .LBB1_4
  cmp rdi, 2
  jne .LBB1_2
...
```

Further, I found that for scenarios where there are more branches. The generated compare instructions should be less than the number of branches if the default branch is undefined behavior. There will be fewer compare instructions if there are some common branches.

Revert the result of #77856: https://llvm-compile-time-tracker.com/compare.php?from=f3c5278efa3b783ada9e7a34b751cf4c5b864535&to=58622ef6755a02f97e5127bea29ed5b8812fe25e&stat=instructions:u.
New change: https://llvm-compile-time-tracker.com/compare.php?from=58622ef6755a02f97e5127bea29ed5b8812fe25e&to=dc2c2faa82d3d7b998680267a79895eb4969e6fd&stat=instructions%3Au.

### CODE REVIEW

**Comment by topperc** [comment]
> But I don't know why it's 4 * InstrCost and not 3 * InstrCost.

On most targets other than X86 load+jump is 2 instructions.

**Comment by DianQK** [comment]
> > But I don't know why it's 4 * InstrCost and not 3 * InstrCost.
> 
> On most targets other than X86 load+jump is 2 instructions.

Thanks. I found that different targets can be quite different in different switch scene. But I think the patch will be at least a little closer now.

**Comment by dtcxzyw** [comment]
Ping for review :) I am happy with the performance data on RISC-V. This patch saves ~8% instructions in some benchmarks.

File: llvm/lib/Analysis/InlineCost.cpp [hunk 1]
```diff
@@ -536,7 +536,13 @@ class CallAnalyzer : public InstVisitor<CallAnalyzer, bool> {
...
 // number of comparisons in a simple closed form :
 //   n + n / 2 - 1 = n * 3 / 2 - 1
-int64_t getExpectedNumberOfCompare(int NumCaseCluster) {
+int64_t getExpectedNumberOfCompare(int NumCaseCluster,
+                                   bool DefaultDestUndefined) {
+  // The compare instruction count should be less than the branch count
+  // when default branch is undefined.
+  if (DefaultDestUndefined) {
+    return static_cast<int64_t>(NumCaseCluster) - 1;
+  }
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/InlineCost.cpp:545]:
```suggestion
  if (DefaultDestUndefined)
    return static_cast<int64_t>(NumCaseCluster) - 1;
```
Please drop the braces.

Is this heuristic correct?

**Comment by DianQK** [comment]
> Ping for review :) I am happy with the performance data on RISC-V. This patch saves ~8% instructions in some benchmarks.

Could you test from d69f4bc19bc04ed0b3c056938b711ea7d7712ef0 to 01838cc6223ceb3c1e0f675865d30aea32db3f71? 58622ef6755a02f97e5127bea29ed5b8812fe25e is just a revert commit.

File: llvm/lib/Analysis/InlineCost.cpp (hunk 1 above)
Comment (DianQK) [reply llvm/lib/Analysis/InlineCost.cpp:545]: Here is an example: https://llvm.godbolt.org/z/x6ETdfY79. If there are common target branches, the number of compare instructions will decrease. I haven't started learning about instruction selection, so I will leave a note here.

**Comment by DianQK** [comment]
Ping. Since the judgment in `getExpectedNumberOfCompare` is still unclear, I have removed this part of the code.

File: llvm/test/Transforms/Inline/inline-switch-default-2.ll [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
-; RUN: opt %s -S -passes=inline -inline-threshold=21 | FileCheck %s
+; RUN: opt %s -S -passes=inline -inline-threshold=11 | FileCheck %s
```
Comment (efriedma-quic) [inline llvm/test/Transforms/Inline/inline-switch-default-2.ll:2]: Please fix the FileCheck lines in this file; there's a bunch of checks for stuff like "LOOKUPTABLE" which don't correspond to a RUN line.  (Maybe push this separately.)

**Review by efriedma-quic — COMMENTED** [review]
Please don't put godbolt links in source code.  If you absolutely need to refer to something, please file a bug and refer to that.

I'd like to see a test that actually verifies the computed cost here.

File: llvm/test/Transforms/Inline/inline-switch-default-2.ll (hunk 2 above)
Comment (DianQK) [reply llvm/test/Transforms/Inline/inline-switch-default-2.ll:2]: Updated: https://github.com/llvm/llvm-project/commit/971ec1f0eea324d4a1eec6709e2c97e1798a6002

**Comment by DianQK** [comment]
> Please don't put godbolt links in source code. If you absolutely need to refer to something, please file a bug and refer to that.

The new issue is https://github.com/llvm/llvm-project/issues/90929.

> I'd like to see a test that actually verifies the computed cost here.

Done. I'm sorry for the late response.

**Review by arsenm — APPROVED** [review]
Seems reasonable to me

**Pull request merged**